### ISSUE TITLE
WIP: Research: Constrained runtime upgrading.

### ISSUE THREAD

**Issue by bedeho** [Joystream/joystream#335]
# WIP

TLDR: How can we put constraints on runtime upgrades in order to preserve some sort of guarantees about what is being changed? e.g. 

- we may want to not do a wholesale upgrading of the entire runtime at a time, perhaps just modules
- even for just a single module, we may want to only upgrade som subcomponent
- we may want to put some constraint on the change that can be made, e.g. block any upgrade which changes the allocation of funds in the platform token.

There is initial work on these topics in the following
- [Cap9/BeakerOS](https://cap9.io/): Cap9 is a capability-based exokernel protocol to run on Ethereum. As a smart contract framework, Cap9 provides developers the ability to perform upgrades in a secure and robust manner - preventing privilege escalation at any point in the development process. 
- [ZeppelinOS](https://zeppelinos.org/): ZeppelinOS is an open-source, distributed platform of tools and services on top of the EVM to develop and manage smart contract applications securely.
- [AragonOS](https://hack.aragon.org/docs/aragonos-intro.html): A computer operating system manages how applications access the underlying resources of the computer—the hardware. aragonOS does the same for decentralized organizations or protocols. It abstracts away the management of how apps and external entities access the resources of the organization or protocol. These resources can be things such as assets, cryptocurrencies, the rights to claim a premium on a loan, or the rights to upgrade a smart contract. Its architecture is based on the idea of a decentralized organization or protocol being the aggregate of multiple components (called applications) connected by a pillar, called the Kernel, which are all governed by a special Access Control List (ACL) application that controls how these applications and other entities can interact with each other.
